src/v/datalake/translation/scheduling.cc

```diff
@@ -0,0 +1,400 @@
+        return;
+    }
+    vlog(datalake_log.trace, "Marking the translator ready: {}", translator);
+    _state.waiting.push_back(translator);
+    _state_changed_cvar.signal();
+}
+
+void scheduler::notify_done(const translator_id& id) {
+    auto holder = _state.gate.hold();
+    vlog(datalake_log.trace, "done notification from translator: {}", id);
+    auto it = _state.translators.find(id);
+    if (it == _state.translators.end()) {
+        return;
+    }
+    auto& translator = it->second;
+    if (
+      translator._waiting_hook.is_linked()
+      || !translator._running_hook.is_linked()) {
+        vlog(
+          datalake_log.warn,
+          "Invalid done notification from waiting translator, ignoring",
+          translator);
+        return;
+    }
+    vlog(datalake_log.trace, "Marking the translator done: {}", translator);
+    translator._running_hook.unlink();
+    translator._stop_in_progress = false;
+    translator._current_translation_start_time.reset();
+    _state_changed_cvar.signal();
+}
+
+bool scheduler::requires_scheduling_actions() const {
+    return !_state.waiting.empty() || _mem_tracker->memory_exhausted();
+}
+
+void scheduler::notify_memory_exhausted() {
+    vlog(datalake_log.debug, "memory exhausted notification");
+    _state_changed_cvar.signal();
+}
+
+ss::future<> scheduler::stop() {
+    vlog(datalake_log.debug, "Stopping scheduler");
+    _state.as.request_abort();
+    _state_changed_cvar.broken();
+    co_await _state.gate.close();
+    co_await ss::max_concurrent_for_each(
+      _state.translators, 32, [](auto& it) mutable {
+          return it.second._translator->stop();
+      });
+}
+
+ss::future<bool>
+scheduler::add_translator(std::unique_ptr<translator> translator) {
+    auto holder = _state.gate.hold();
+    const auto& id = translator->id();
+    const auto& translators = _state.translators;
+    vlog(
+      datalake_log.trace,
+      "request to add translator with id: {}, current_translators: {}",
+      id,
+      translators.size());
+    auto it = translators.find(id);
+    if (translators.contains(id)) {
+        vlog(
+          datalake_log.error,
+          "duplicate translator registration: {}",
+          it->second);
+        co_return false;
+    }
+    _state.translators[id] = translator_state{std::move(translator)};
+    co_await _state.translators[id]._translator->start(*this, *_mem_tracker);
+    co_return true;
+}
+
+ss::future<> scheduler::remove_translator(const translator_id& id) {
+    auto holder = _state.gate.hold();
+    vlog(
+      datalake_log.trace,
+      "request to remove translator with id: {}, current_translators: {}",
+      id,
+      _state.translators.size());
+    auto it = _state.translators.find(id);
+    if (it == _state.translators.end()) {
+        co_return;
+    }
+    it->second._waiting_hook.unlink();
+    it->second._running_hook.unlink();
+    auto translator = std::move(it->second);
+    _state.translators.erase(it);
```

Comment:
Do we need to worry about other components holding on to this across scheduling points and it being deleted/destroyed while they are use it?

I guess the current implementation is careful not to do that, but I'm slightly worried about not accidently doing something like this in the future. Can we at the very least add some documentation to `scheduler_state::translators` that it's not safe to hold a translator across a scheduling point.

Reply:
Thats a good point, I added a note.  I thought that was more or less implicit given seastar restrictions but agreed that it will be good to document in the header.

One thing to keep in mind about this `translators` map is that it is used only by the "scheduler" but not by the "scheduling_policy" (although there is no enforcement for it). Scheduling policy only looks at running/waiting lists but the same rule applies to those lists as well (be mindful of scheduling points).